_posts/2020-12-28-gcp-point-in-time.md

```diff
@@ -0,0 +1,98 @@
+---
+layout: post
+title: Restoring CloudSQL from a point-in-time backup
+subtitle: Fun with MySQL binlogs
+tags: [binlog, clousql, mysql]
+image: /assets/img/mysql.png
+comments: true
+---
+
+We've had a problem. We need to restore our CloudSQL MySQL database using point-in-time recovery. There's a runbook of how to do that, right?
+This is a technical write up of the next few hours of learning how to actually do this in a sane way.
+
+# How did we get here?
+
+At GoCardless, we have a service that runs on a CloudSQL powered MySQL instance. Something had *gone wrong* and we needed to recover some deleted data, ideally 'just before' it was deleted. GCP provides two backup types: a scheduled one and a 'point in time' one. We tried restoring from one of the scheduled daily backups as a starting point, but that failed instantly. Instead, we started looking at 'point in time recovery'. To do this, we needed to provide a location in the binary log file.
+
+# A binary log file, you say?
+
+There's a detailed overview in the [MySQL Docs](https://dev.mysql.com/doc/internals/en/binary-log-overview.html#:~:text=The%20binary%20log%20is%20a,was%20introduced%20in%20MySQL%203.23.), but the TLDR is:
+
+The binary log is a set of log files that contains a list of statements which could have updated data in the MySQL instance (it ignores reads). It includes a timestamp, a duration (how long the query took to complete) and some internal information that MySQL uses to do its thing. It's used by the master replica to send events to any secondary servers to keep them up to date (via something called a relay log), but it can also be used for data recovery (i.e. to replay everything that happened to a database up until a particular point).
+
+
+# This is all sounding sensible
+
+So the fun part here is 'how do I get from a timestamp to a binary log file location'. We knew that we wanted to restore to just before 1:02pm - we called it 1pm to be on the safe side. Some StackOverflow-ing later, we got to the following set of steps.
+
+### 1. mysql-server
+
+To do anything useful with the binary logs, you need to be able to parse them. MySQL provides a nice utility for this: [`mysqlbinlog`](https://dev.mysql.com/doc/refman/8.0/en/mysqlbinlog.html). Understandably, it's packaged inside the CLI utility `mysql-server`. Our application pods (like most such pods, I would expect) have `mysql-client` installed by not `mysql-server`. So step 1 was adding a line to our application pod Dockerfile
+```
+apt-get install mysql-server
+```
+
+Note that it may have been possible to SSH onto the CloudSQL pod itself, but we didn't have the permissions to do that so we chose to use the application pods instead.
```

Comment:
i don't think you can do this. you mean the virtual machine running the actual database server? Pretty sure you can't do that, i'd kill this line.